airflow/models/dagrun.py

```diff
@@ -593,17 +604,22 @@ def recalculate(self) -> _UnfinishedStates:
                     msg='task_failure',
                 )
 
+                if notification:
+                    notification(dag_run=self, callback_request=cb)
+                if dag.has_on_success_callback:
```

Comment:
maybe better would be to not combine the notification into this if / else block in the way you've done ...  the nested branching could be more complicated than necessary?   e.g. case in point here this is supposed to be `has_on_success_callback` i think.  i'm not 100% sure there's a better way, but there _could_ be and it might be worth giving it a  think

Reply:
Changed that, also the notification now does not rely on DagCallbackRequest - since we don't always have it, like in BackfillJob.